vllm/model_executor/models/molmo.py

```diff
@@ -0,0 +1,1130 @@
+            positions=positions,
+            kv_caches=kv_caches,
+            attn_metadata=attn_metadata,
+            inputs_embeds=inputs_embeds,
+        )
+
+        return hidden_states
+         
+    def compute_logits(self, hidden_states: torch.Tensor,
+                       sampling_metadata: SamplingMetadata) -> torch.Tensor:
+        logits = self.logits_processor(self.lm_head, hidden_states,
+                                       sampling_metadata)
+        return logits
+
+    def sample(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata:  SamplingMetadata,
+    ) -> Optional[SamplerOutput]:
+        next_tokens = self.sampler(logits, sampling_metadata)
+        return next_tokens
+
+    def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
+
+        params_dict = dict(self.named_parameters(remove_duplicate=False))
+
+        embedding_weight = dict()
+        projector_weight = dict()
+        for name, loaded_weight in weights:
+            log.info(f"Original name: {name}")
+            if "rotary_emb.inv_freq" in name:
+                log.info(f"Skipping {name}")
+                continue
+            if self.config.tie_word_embeddings and "lm_head.weight" in name:
+                continue
+
+            if "wte.embedding" in name:
+                embedding_weight["embedding"] = loaded_weight
+                continue
+                
+            if "wte.new_embedding" in name:
+                embedding_weight["new_embedding"] = loaded_weight
+                continue
+           
+            if "vision_backbone" in name:
+                if name.startswith("model"):
+                    name = name[len("model."):]
+                if 'image_projector' in name:
+                    if 'w1' in name:
+                        projector_weight['gate_proj'] = loaded_weight
+                    elif 'w3' in name:
+                        projector_weight['up_proj'] = loaded_weight
+                    elif 'w2' in name:
+                        projector_weight['down_proj'] = loaded_weight
+                    else:
+                        raise ValueError(f"Unexpected projector weight: {name}")
+                    continue
+            else:
+                if "ln_f.weight" in name:
+                    name = "model.norm.weight"
+
+                if "transformer.blocks" in name:
+                    name = name.replace("transformer.blocks", "layers")
+    
+                if "attn_out" in name:
+                    name = name.replace("attn_out", "self_attn.o_proj")
+                
+                if "att_proj" in name:
+                    name = name.replace("att_proj", "self_attn.qkv_proj")
+                
+                if 'q_norm' in name:
+                    name = name.replace("q_norm", "self_attn.q_norm")
+                
+                if 'k_norm' in name:
+                    name = name.replace("k_norm", "self_attn.k_norm")
```

Comment:
Can you use a mapping for this?

Reply:
Apologies for the delay in the update. I believe I have completed the update except for the weight loading and the reuse of `CLIPParallelAttention`. I have two questions:
1. Could you point me to the reference code using the "mapping"? It seems that mapping is also required to use CLIPParallelAttention.
2. After updating the code, should I create a new PR separate from this one?